fvm/evm/offchain/sync/replay.go

```diff
@@ -0,0 +1,250 @@
+	ctx.TxCountSoFar = txIndex
+	// populate precompiled calls
+	if len(txEvent.PrecompiledCalls) > 0 {
+		pcs, err := types.AggregatedPrecompileCallsFromEncoded(txEvent.PrecompiledCalls)
+		if err != nil {
+			return fmt.Errorf("error decoding precompiled calls [%x]: %w", txEvent.Payload, err)
+		}
+		ctx.ExtraPrecompiledContracts = precompiles.AggregatedPrecompiledCallsToPrecompiledContracts(pcs)
+	}
+
+	// create a new block view
+	bv, err := em.NewBlockView(ctx)
+	if err != nil {
+		return err
+	}
+
+	var res *types.Result
+	// check if the transaction payload is actually from a direct call,
+	// which is a special state transition in Flow EVM.
+	if txEvent.TransactionType == types.DirectCallTxType {
+		call, err := types.DirectCallFromEncoded(txEvent.Payload)
+		if err != nil {
+			return fmt.Errorf("failed to RLP-decode direct call [%x]: %w", txEvent.Payload, err)
+		}
+		res, err = bv.DirectCall(call)
+		if err != nil {
+			return fmt.Errorf("failed to execute direct call [%x]: %w", txEvent.Hash, err)
+		}
+	} else {
+		gethTx := &gethTypes.Transaction{}
+		if err := gethTx.UnmarshalBinary(txEvent.Payload); err != nil {
+			return fmt.Errorf("failed to RLP-decode transaction [%x]: %w", txEvent.Payload, err)
+		}
+		res, err = bv.RunTransaction(gethTx)
+		if err != nil {
+			return fmt.Errorf("failed to run transaction [%x]: %w", txEvent.Hash, err)
+		}
+	}
+
+	// validate results
+	if validate {
+		if err := ValidateResult(res, txEvent); err != nil {
+			return fmt.Errorf("transaction replay failed (txHash %x): %w", txEvent.Hash, err)
+		}
+	}
+
+	return nil
+}
+
+func ValidateResult(
+	res *types.Result,
+	txEvent *events.TransactionEventPayload,
+) error {
+
+	// we should never produce invalid transaction, since if the transaction was emitted from the evm core
+	// it must have either been successful or failed, invalid transactions are not emitted
+	if res.Invalid() {
+		return fmt.Errorf("invalid transaction: %w", res.ValidationError)
+	}
+
+	// check gas consumed
+	if res.GasConsumed != txEvent.GasConsumed {
+		return fmt.Errorf("gas consumption mismatch %d != %d", res.GasConsumed, txEvent.GasConsumed)
+	}
+
+	// check error msg
+	if errMsg := res.ErrorMsg(); errMsg != txEvent.ErrorMessage {
+		return fmt.Errorf("error msg mismatch %s != %s", errMsg, txEvent.ErrorMessage)
+	}
+
+	// check encoded logs
+	encodedLogs, err := res.RLPEncodedLogs()
+	if err != nil {
+		return fmt.Errorf("failed to RLP-encode logs: %w", err)
+	}
+	if !bytes.Equal(encodedLogs, txEvent.Logs) {
+		return fmt.Errorf("encoded logs mismatch %s != %s", encodedLogs, txEvent.ErrorMessage)
+	}
+
+	// check deployed address
+	if deployedAddress := res.DeployedContractAddressString(); deployedAddress != txEvent.ContractAddress {
+		return fmt.Errorf("deployed address mismatch %s != %s", deployedAddress, txEvent.ContractAddress)
+	}
+
+	// check the state change checksum
+	// if empty checksum skip (supporting blocks before checksum integration)
```

Comment:
can we remove the skip at some point?

Reply:
could you change this check to check on `txEvent.BlockHeight` instead of `txEvent.StateUpdateChecksum`?

Reply:
I kept it this way, since we plan to disable checksum (leave it empty) sometime in the future when everything is stable and we are confident about the re-player, so we get some throughput gain on the ENs. that way we don't need changes to this package